Add support for setting RGB and RGBW values for Twinkly lights

RobBie1221 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Change library from twinkly-client to ttls. It covers same functionality and more and has required functions to set RGBW values.
Add support in the light entity for setting RGBW values.
Update all tests.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [X] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [X] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

RobBie1221 committed:
Change library to ttls

RobBie1221 committed:
Add rgbw support

RobBie1221 committed:
Add client session to config flow

RobBie1221 committed:
Fix config flow

RobBie1221 committed:
Adjust tests 1

RobBie1221 committed:
Fix more tests

RobBie1221 committed:
Fix last tests

RobBie1221 committed:
Add new tests

RobBie1221 committed:
Update test for coverage

RobBie1221 committed:
Update test for coverage 2

RobBie1221 committed:
Update test for coverage 3

pvizeli commented:
Nice. Is also possible to read effect and let them choose with the library?

RobBie1221 commented:
Maybe...to set fixed color mode "movie" is used with 1 single frame with a fixed color. There is a mode "effect", but I didn't test that yet.

I did forget one thing by the way when submitting this PR, there is twinkly RGB, RGBW and AWW. I need to add that it reads the available mode and set color_mode accordingly.

dr1rrb commented:
> Change library from twinkly-client to ttls. It covers same functionality and more and has required functions to set RGBW values.
> Add support in the light entity for setting RGBW values.

@RobBie1221 😍😍😍

> Nice. Is also possible to read effect and let them choose with the library?

@pvizeli  From my last investigations last year, unfortunately not really: effects are not stored on the device, they are sent by the app each time. And when you "group" and "map" multiple devices together, effects are "computed" by the app before being sent, so it's not a mater of capturing all of them statically and embed them in the component.

The single option I see would be to have a "listening mode" which would somehow act as a "an in the middle" to capture effects sent by the app, store them on HA and then expose them as "light effects" in HA ...

dr1rrb commented:
Hey @RobBie1221 does the `ttls.client` supports the "Playlist" mode? It's a feature I completely missed by the time which would be great to support (cf. https://github.com/dr1rrb/py-twinkly-client/issues/1).

(BTW when ready I do have RGB and RGBW devices, I will be more than happy to test it)

RobBie1221 commented:
> Hey @RobBie1221 does the `ttls.client` supports the "Playlist" mode? It's a feature I completely missed by the time which would be great to support (cf. [dr1rrb/py-twinkly-client#1](https://github.com/dr1rrb/py-twinkly-client/issues/1) and #62370).

Don't know, need to check that.

> (BTW when ready I do have RGB and RGBW devices, I will be more than happy to test it)

I do have RGBW, but don't have RGB, so it would be good to have someone who can test that.

matejdro commented:
Yeah, twinkly integration with home assistant presents at least two difficult challenges:

1. There is no way to enter `playlist` mode. Twinkly lights have (at least) three modes of operation: `movie`, `playlist` and `off`. When you just select an effect from the app, it switches to the `movie` mode. But when you start a playlist, it switches to `playlist` mode. However, turning lights off and on in home assistant will first change its mode to `off` [and then back to `movie`](https://github.com/jschlyter/ttls/blob/main/ttls/client.py#L261), which disables the playlist. I'm not sure what would be a good way to map these three modes to two-state home assistant switches. Maybe some kind of config option or parameter on `turn_on` service?

2. Twinkly lights can only be connected to one client at a time. When a client connects, it kicks all others out. This presents a problem when attempting to upload a playlist onto device from the app. This process is very slow, so it can take up to a minute. But home assistant polls more frequently than that, killing your upload. It's essentially impossible to upload any nontrivial playlist when home assistant entity being active, since its polling keep disconnecting the app.

I'm not saying these should be resolved right now (this PR has already improved twinkly situation significantly as it resolves the brightness changing killing playlist fix), I just want to bring them to your attention since we appear to have a new Twinkly maintainer.

dr1rrb commented:
> Yeah, twinkly integration with home assistant presents at least two difficult challenges:
> 
>     1. [...] Maybe some kind of config option for default mode and/or parameter on `turn_on` service?

Great idea

> 
>     2. Twinkly lights can only be connected to one client at a time. When a client connects, it kicks all others out. This presents a problem when attempting to upload a playlist onto device from the app. This process is very slow, so it can take up to a minute. But home assistant polls more frequently than that, killing your upload. It's essentially impossible to upload any nontrivial playlist when home assistant entity is active, since its polling keep disconnecting the app. Maybe we could work around that by adding a switch to twinkly lights that disables polling?
> 

My idea for this was to add a basic flag/switch to disable the polling for a given time (like 20 min). Then we could also add some basic logic like when we get a 401 we automatically enable that flag, and when we have a "real" command from HA we auto-disable it.

But this year there is a new setting in the Twinkly App which sounds promising: you can make your device discoverable to homekit. I would assume that we would be able to avoid the "auth conflict" issue.

wicol commented:
> Maybe...to set fixed color mode "movie" is used with 1 single frame with a fixed color.

Can I ask why it'd be preferable doing this over using the "color" mode?

RobBie1221 commented:
> > Maybe...to set fixed color mode "movie" is used with 1 single frame with a fixed color.
> 
> Can I ask why it'd be preferable doing this over using the "color" mode?

What color mode? These I know off:

https://github.com/jschlyter/ttls/blob/707e860109f7aacc613e8d9aa98038498d9e37ff/ttls/client.py#L47

wicol commented:
> What color mode? These I know off:
> 
> https://github.com/jschlyter/ttls/blob/707e860109f7aacc613e8d9aa98038498d9e37ff/ttls/client.py#L47

I got it from here:
https://xled-docs.readthedocs.io/en/latest/rest_api.html#get-led-operation-mode
https://xled-docs.readthedocs.io/en/latest/rest_api.html#set-led-color

I used these successfully with my RGBW and ended up creating a [PR ](https://github.com/dr1rrb/py-twinkly-client/pull/4)for twinkly-client. Then I found this PR haha "oops".

RobBie1221 commented:
I think that API is new. Anyway it would take major redesign to get it working like that. The integration works at the moment that "on" means set mode "movie". If we are going to support multiple modes, we need a design to incorporate that.

The simplicity of this solution is that nothing changes to which modes are set. I think in the end it may be nicer to use the "color" mode, but that needs more work. Also it needs to be in a library. Perhaps you can make a PR against the ttls library, that one seems to get regular updates.

wicol commented:
> I think that API is new. Anyway it would take major redesign to get it working like that. The integration works at the moment that "on" means set mode "movie". If we are going to support multiple modes, we need a design to incorporate that.

I see. I just got my twinkly last week, so that's when I started to look around for options and libs. I saw some ppl playing with xled and xled_plus, and then focused on twinkly-client since that's the one used by HA. I completely missed ttls though 🙄 

> The simplicity of this solution is that nothing changes to which modes are set. I think in the end it may be nicer to use the "color" mode, but that needs more work. Also it needs to be in a library. Perhaps you can make a PR against the ttls library, that one seems to get regular updates.

I agree, this should get merged, given that it's a major improvement over the current implementation and ppl (including myself) want this stuff asap 😄  . Meanwhile I'll try to play around with ttls and see if I can add color mode to it 👍

pvizeli commented on homeassistant/components/twinkly/light.py:
```diff
@@ -53,6 +54,10 @@ def __init__(
         self._hass = hass
         self._conf = conf
 
+        self._attr_supported_color_modes = {COLOR_MODE_RGBW}
+        self._attr_color_mode = COLOR_MODE_RGBW
+        self._attr_rgbw_color = (255, 255, 255, 0)
```
Well not all support RGBW, we should handle it dynamic and only show the support if they are present over an property

RobBie1221 replied:
Yep, you're right. It got burried a bit over the numerous comments. It's on my list to adjust.

pvizeli commented:
All comments have to stop now, by side of the code review. Feel free to open a forum topic or an issue for future discusion.

pvizeli commented on homeassistant/components/twinkly/light.py:
```diff
@@ -148,31 +148,50 @@ async def async_turn_on(self, **kwargs) -> None:
             )
 
-            device_info = await self._client.get_device_info()
+            self._brightness = int(round(brightness_value * 2.55)) if self._is_on else 0
```
Just set `https://github.com/home-assistant/core/blob/dev/homeassistant/components/light/__init__.py#L692`, not need to have an property above

pvizeli reviewed (commented):
You can also remove line 132 - 141 - That is wrong `extra_state_attributes`

RobBie1221 commented:
> You can also remove line 132 - 141 - That is wrong `extra_state_attributes`

I noticed that there is still a lot of review comments open on the initial PR for the integration, which I think also adresses this. Maybe make a separate PR to cleanup those comments?

RobBie1221 committed:
Change brightness to attribute

RobBie1221 commented:
> You can also remove line 132 - 141 - That is wrong `extra_state_attributes`

@pvizeli No problem with removing that, but isn't that a breaking change for people using that? Shouldn't we do that in a separate pull?

RobBie1221 committed:
Set RGBW mode only when available

RobBie1221 committed:
Add RGB support

RobBie1221 commented:
I reworked it a bit:
- It now defaults to COLOR_MODE_BRIGHTNESS. For devices reporting rgb or rgbw, it will change color mode to the matching mode.
- I cleaned up some bits and pieces as mentioned here: https://github.com/home-assistant/core/pull/42103#pullrequestreview-534858588
- Brightness now uses attribute, but there is more cleaning up to do (I would propose in a follow-up PR, not to make this any bigger than it's already is)

pvizeli commented:
Looks a lot better as that code before :D

MartinHjelmare commented on homeassistant/components/twinkly/__init__.py:
```diff
@@ -1,28 +1,41 @@
+    hass.data[DOMAIN][uuid][DATA_CLIENT] = client
+    hass.data[DOMAIN][uuid][DATA_DEVICE_INFO] = device_info
 
     hass.config_entries.async_setup_platforms(entry, PLATFORMS)
```
Side note: I noticed that we don't forward the entry unload to the platforms when unloading the config entry below. Would be good to fix that in a follow up.

MartinHjelmare commented on homeassistant/components/twinkly/__init__.py:
```diff
@@ -1,28 +1,41 @@
-    hass.data.setdefault(DOMAIN, {})[uuid] = twinkly_client.TwinklyClient(
-        host, async_get_clientsession(hass)
-    )
+    hass.data[DOMAIN].setdefault(uuid, {})
```
Our standard is to use the config entry entry_id as key under the domain key in hass.data.

MartinHjelmare commented on homeassistant/components/twinkly/light.py:
```diff
@@ -35,7 +45,10 @@ async def async_setup_entry(
     """Setups an entity from a config entry (UI config flow)."""
 
-    entity = TwinklyLight(config_entry, hass)
+    client = hass.data[DOMAIN][config_entry.data[CONF_ENTRY_ID]][DATA_CLIENT]
```
Would be good to rename `CONF_ENTRY_ID` to something that is less mistakable for config entry entry_id.

MartinHjelmare commented on homeassistant/components/twinkly/light.py:
```diff
@@ -136,7 +148,7 @@ def extra_state_attributes(self) -> dict:
         # Make sure to update any normalized property
         attributes[ATTR_HOST] = self._client.host
-        attributes[ATTR_BRIGHTNESS] = self._brightness
+        attributes[ATTR_BRIGHTNESS] = self._attr_brightness
```
We should just remove the extra state attributes here. Brightness is set by the parent light class in the state attributes, and we don't want static info like host in the state attributes.

MartinHjelmare commented on tests/components/twinkly/test_twinkly.py:
```diff
@@ -10,7 +10,6 @@
     CONF_ENTRY_NAME,
     DOMAIN as TWINKLY_DOMAIN,
 )
```
We should rename this test module to `test_light.py`.

MartinHjelmare reviewed (commented):
Nice! All my comments pertain to code that was already there from earlier contributions. Just noting them again, to help with possible future clean up efforts. ❤️

RobBie1221 commented:
I'll make a follow up PR :)
